hass-dash.el

```diff
@@ -143,193 +91,136 @@
   :group 'hass-dash
   :type 'string)
 
-(defcustom hass-dash-layout nil
-  "A list of widgets to show on the dashboard.
-Each element in the `list' is an `alist' of a Group name to a `plist' of entity
-IDs with their properties.
-
-The `car' of a list is the group label while the `cdr' is a list of widget
-definitions for that group.
-
-'((\"Group Label\" . ((\"entity.id_example\" :label \"Human Readable Name\"))))
-
-See `hass-dash--create-widget' for widget properties.
+(defvar hass-dash-layout nil
+  "A list describing the root widget to show on the dashboard.
+In most cases, this will be a `group' widget.  You can then build a tree of
+arbitrary widgets to display on the dashboard.  You'll probably want to make use
+of hass widgets such as `hass-dash-toggle' or `hass-dash-group'.
 
 Full example:
 
-\(setq `hass-dash-layout'
- '((\"Group One\" . ((\"input_boolean.test_boolean\"
-                    :label \"Toggle entity\")
-                   (\"switch.bedroom_light\"
-                    :label \"Bedroom Light\")
-                   (\"input_boolean.test_boolean\"
-                    :label \"Turn off test boolean\"
-                    :service \"input_boolean.turn_off\")
-                   (\"automation.some_automation\")))
-   (\"Vacuum Group\" . ((\"vacuum.valetudo_vacuum\"
-                       :label \"Vacuum\")
-                      (\"vacuum.valetudo_vacuum\"
-                       :label \"Vacuum return home\"
-                       :service \"vacuum.return_to_base\"
-                       :state nil
-                       :icon nil)))))"
-  :group 'hass-dash
-  :type 'list)
-
-
-;; Default formatters
-(defcustom hass-dash-default-widget-formatter #'hass-dash-widget-formatter
-  "The function called to format the widgets on the dashboard."
-  :group 'hass-dash
-  :type 'function)
-
-(defcustom hass-dash-default-label-formatter #'hass-dash-label-formatter
-  "The function called to format the label of widgets on the dashboard."
-  :group 'hass-dash
-  :type 'function)
-
-(defcustom hass-dash-default-state-formatter #'hass-dash-state-formatter
-  "The function called to format the state of widgets on the dashboard."
-  :group 'hass-dash
-  :type 'function)
-
-(defcustom hass-dash-default-icon-formatter #'hass-dash-icon-formatter
-  "The function called to format the icon of widgets on the dashboard."
-  :group 'hass-dash
-  :type 'function)
+\(setq hass-dash-layout
+      \\=`(hass-dash-group
+        :title \"Home Assistant\"
+        :format \"%t\\n\\n%v\"
+        (hass-dash-group
+         :title \"Kitchen\"
+         :title-face outline-2
+         (hass-dash-toggle :entity-id \"light.kitchen_lights\")
+         (hass-dash-toggle :entity-id \"switch.entry_light\"
+                           :label \"Hallway\"
+                           :confirm t))))")
```

Comment:
I suppose the reason you went with creating a single top-level group is to support different layouts. Totally cool with me. I think it would be nice though if the groups (particularly the top-level one) didn't require a title.

An easy solution for the user is just setting `:format "%v"`, but then they also have to set `:title` to something like an empty string even though it's not used.

Not sure if you have a better solution, but I feel that a title-less group might be a common enough need that it should be explicitly documented at the very least.

Reply:
For that you can just use the built-in `group` widget (actually the `hass-dash-group` widget is just a wrapper around the `group` widget but adding a title with a face).